docs/docs/noir/concepts/comptime.md

```diff
@@ -0,0 +1,358 @@
+sidebar_position: 15
+---
+
+# Overview
+
+Metaprogramming in Noir is comprised of three parts:
+1. `comptime` code
+2. Quoting and unquoting
+3. The metaprogramming API in `std::meta`
+
+Each of these are explained in more detail in the next sections but the wide picture is that
+`comptime` allows us to write code which runs at compile-time. In this `comptime` code we
+can quote and unquote snippets of the program, manipulate them, and insert them in other
+parts of the program. Comptime functions which do this are said to be macros. Additionally,
+there's a compile-time API of built-in types and functions provided by the compiler which allows
+for greater analysis and modification of programs.
+
+---
+
+# Comptime
+
+`comptime` is a new keyword in Noir which marks an item as executing or existing at compile-time. It can be used in several ways:
+
+- `comptime fn` to define functions which execute exclusively during compile-time.
+- `comptime global` to define a global variable which is evaluated at compile-time.
+  - Unlike runtime globals, `comptime global`s can be mutable.
+- `comptime { ... }` to execute a block of statements during compile-time.
+- `comptime let` to define a variable whose value is evaluated at compile-time.
+- `comptime for` to run a for loop at compile-time. Syntax sugar for `comptime { for .. }`.
+
+## Scoping
+
+Note that while in a `comptime` context, any runtime variables _local to the current function_ are never visible.
+
+## Evaluating
+
+Evaluation rules of `comptime` follows the normal evaluation rules for other Noir code. There are a few things to note though:
+
+- Certain built-in functions may not be available, although more may be added over time.
+- Evaluation order of global items is currently unspecified. For example, given the following two functions we can't guarantee
+which `println` will execute first. The ordering of the two printouts will be arbitrary, but should be stable across multiple compilations as long as the program is also unchanged.
+
+```rs=
+fn one() {
+    comptime { println("one"); }
+}
+
+fn two() {
+    comptime { println("two"); }
+}
+```
+
+- Since evaluation order is unspecified, care should be taken when using mutable globals so that they do not rely on a particular ordering.
+For example, using globals to generate unique ids should be fine but relying on certain ids always being produced (especially after edits to the program) should be avoided.
+- Although most ordering of globals is unspecified, two are:
+  - Dependencies of a crate will always be evaluated before the dependent crate.
+  - `comptime` items are evaluated before runtime items, however `comptime` functions can call runtime functions in which case
+    the runtime functions will be executed at compile-time.
+
+## Lowering
+
+When a `comptime` value is used in runtime code it must be lowered into a runtime value. This often means replacing the expression with the literal that it evaluated to. For example, the code:
```

Comment:
`This often means replacing the expression with the literal that it evaluated to.`
What else does it mean?

Reply:
Quoted code will be lowered into each token then parsed as an expression for example. Although I think I'm going to move to calling that macro insertion or something since you can also lower quoted code down to an actual `quote { ... }` expression.

Reply:
Reworded/clarified this and the next section once we talk about Quoted values